From 9aeab55d09f9df833bca4467b0a209cea2901ede Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Tim=20R=C3=BChsen?= <tim.ruehsen@gmx.de>
Date: Thu, 28 Mar 2024 18:12:19 +0100
Subject: [PATCH] Fix --no-parent for denormalized paths

* libwget/iri.c (wget_iri_parse): Normalize path part of URL.
* unit-tests/test.c (test_iri_parse): Add test with denormalized path.
---
 libwget/iri.c     | 3 +++
 unit-tests/test.c | 1 +
 2 files changed, 4 insertions(+)

diff --git a/libwget/iri.c b/libwget/iri.c
index 8241ea971..13bd5259b 100644
--- a/libwget/iri.c
+++ b/libwget/iri.c
@@ -82,6 +82,8 @@ static struct iri_scheme {
 	[WGET_IRI_SCHEME_HTTPS] = { 443, "https" },
 };
 
+static size_t WGET_GCC_NONNULL_ALL normalize_path(char *path);
+
 /**
  * \param[in] scheme Scheme to get name for
  * \return Name of \p scheme (e.g. "http" or "https") or NULL is not supported
@@ -561,6 +563,7 @@ wget_iri *wget_iri_parse(const char *url, const char *encoding)
 		c = *s;
 		if (c) *s++ = 0;
 		wget_iri_unescape_inline((char *)iri->path);
+		normalize_path((char *)iri->path);
 	}
 
 	if (c == '?') {
diff --git a/unit-tests/test.c b/unit-tests/test.c
index da8cc728b..80ddeced5 100644
--- a/unit-tests/test.c
+++ b/unit-tests/test.c
@@ -584,6 +584,7 @@ static void test_iri_parse(void)
 		{ "http://example+.com/pa+th?qu+ery#fr+ag", NULL, WGET_IRI_SCHEME_HTTP, NULL, NULL, "example+.com", 80, "pa+th", "qu ery", "fr+ag"},
 		{ "http://example.com#frag?x", NULL, WGET_IRI_SCHEME_HTTP, NULL, NULL, "example.com", 80, NULL, NULL, "frag?x"},
 		{ "http://user:pw@example.com", NULL, WGET_IRI_SCHEME_HTTP, "user", "pw", "example.com", 80, NULL, NULL, NULL},
+		{ "http://example.com//path//file", NULL, WGET_IRI_SCHEME_HTTP, NULL, NULL, "example.com", 80, "path/file", NULL, NULL},
 	};
 	unsigned it;
 
-- 
GitLab

